File: frontend/src/api/releases.ts
/**
 * API client for the SpecForge Release Manager (Module 5).
 *
 * Endpoints:
 *   GET    /api/v1/projects/{projectId}/releases
 *   POST   /api/v1/projects/{projectId}/releases
 *   GET    /api/v1/projects/{projectId}/releases/changelog
 *   GET    /api/v1/projects/{projectId}/releases/summary
 *   GET    /api/v1/projects/{projectId}/releases/{version}
 *   PUT    /api/v1/projects/{projectId}/releases/{version}
 *   DELETE /api/v1/projects/{projectId}/releases/{version}
 *   POST   /api/v1/projects/{projectId}/releases/{version}/changes
 *   PUT    /api/v1/projects/{projectId}/releases/{version}/status
 *   DELETE /api/v1/projects/{projectId}/releases/{version}/changes/{changeId}
 */

const BASE = '/api/v1';

// ---------------------------------------------------------------------------
// Domain types
// ---------------------------------------------------------------------------

export type ChangeType =
  | 'feat'
  | 'fix'
  | 'docs'
  | 'refactor'
  | 'test'
  | 'chore'
  | 'breaking';

export type ReleaseStatus = 'draft' | 'published' | 'yanked';

export const CHANGE_TYPE_ORDER: ChangeType[] = [
  'breaking',
  'feat',
  'fix',
  'refactor',
  'docs',
  'test',
  'chore',
];

export const CHANGE_TYPE_LABEL: Record<ChangeType, string> = {
  breaking: '💥 Breaking Changes',
  feat: '✨ Features',
  fix: '🐛 Bug Fixes',
  refactor: '♻️ Refactoring',
  docs: '📝 Documentation',
  test: '✅ Tests',
  chore: '🔧 Chores',
};

export const STATUS_COLOR: Record<ReleaseStatus, string> = {
  draft: '#6b7280',
  published: '#059669',
  yanked: '#dc2626',
};

export const STATUS_ICON: Record<ReleaseStatus, string> = {
  draft: '📝',
  published: '🚀',
  yanked: '⛔',
};

export interface ChangeEntry {
  change_id: string;
  change_type: ChangeType;
  description: string;
  task_id: string | null;
  story_id: string | null;
}

export interface ReleaseSchema {
  version: string;
  title: string;
  date: string;
  status: ReleaseStatus;
  task_list_id: string | null;
  spec_id: string | null;
  plan_id: string | null;
  notes: string;
  changes: ChangeEntry[];
}

export interface ReleaseLogResponse {
  project_name: string;
  notes: string;
  releases: ReleaseSchema[];
}

export interface ChangelogResponse {
  markdown: string;
}

export interface ReleaseLogSummary {
  total_releases: number;
  by_status: Record<ReleaseStatus, number>;
  total_changes: number;
  by_change_type: Record<ChangeType, number>;
}

// ---------------------------------------------------------------------------
// HTTP helpers
// ---------------------------------------------------------------------------

async function request<T>(url: string, init?: RequestInit): Promise<T> {
  const res = await fetch(url, {
    headers: { 'Content-Type': 'application/json', ...(init?.headers ?? {}) },
    ...init,
  });
  if (!res.ok) {
    const text = await res.text();
    throw new Error(`HTTP ${res.status}: ${text}`);
  }
  if (res.status === 204) return undefined as unknown as T;
  return res.json() as Promise<T>;
}

// ---------------------------------------------------------------------------
// API functions
// ---------------------------------------------------------------------------

export function getReleaseLog(projectId: string): Promise<ReleaseLogResponse> {
  return request(`${BASE}/projects/${projectId}/releases`);
}

export function createReleaseLog(
  projectId: string,
  body: Partial<ReleaseLogResponse>
): Promise<ReleaseLogResponse> {
  return request(`${BASE}/projects/${projectId}/releases`, {
    method: 'POST',
    body: JSON.stringify(body),
  });
}

export function getChangelog(projectId: string): Promise<ChangelogResponse> {
  return request(`${BASE}/projects/${projectId}/releases/changelog`);
}

export function getReleaseLogSummary(projectId: string): Promise<ReleaseLogSummary> {
  return request(`${BASE}/projects/${projectId}/releases/summary`);
}

export function getRelease(projectId: string, version: string): Promise<ReleaseSchema> {
  return request(`${BASE}/projects/${projectId}/releases/${version}`);
}

export function updateRelease(
  projectId: string,
  version: string,
  body: Partial<ReleaseSchema>
): Promise<ReleaseSchema> {
  return request(`${BASE}/projects/${projectId}/releases/${version}`, {
    method: 'PUT',
    body: JSON.stringify(body),
  });
}

export function deleteRelease(projectId: string, version: string): Promise<void> {
  return request(`${BASE}/projects/${projectId}/releases/${version}`, {
    method: 'DELETE',
  });
}

export function addChangeEntry(
  projectId: string,
  version: string,
  entry: Omit<ChangeEntry, 'change_id'>
): Promise<ChangeEntry> {
  return request(`${BASE}/projects/${projectId}/releases/${version}/changes`, {
    method: 'POST',
    body: JSON.stringify(entry),
  });
}

export function updateReleaseStatus(
  projectId: string,
  version: string,
  status: ReleaseStatus
): Promise<ReleaseSchema> {
  return request(`${BASE}/projects/${projectId}/releases/${version}/status`, {
    method: 'PUT',
    body: JSON.stringify({ status }),
  });
}

export function removeChangeEntry(
  projectId: string,
  version: string,
  changeId: string
): Promise<void> {
  return request(
    `${BASE}/projects/${projectId}/releases/${version}/changes/${changeId}`,
    { method: 'DELETE' }
  );
}

// ---------------------------------------------------------------------------
// Demo data
// ---------------------------------------------------------------------------

export function getDemoReleaseLog(): ReleaseLogResponse {
  return {
    project_name: 'Demo Project',
    notes: 'Auto-generated demo release log for Module 5.',
    releases: [
      {
        version: '5.0.0',
        title: 'Release Manager',
        date: '2026-04-03',
        status: 'published',
        task_list_id: 'tasks-demo',
        spec_id: 'spec-demo',
        plan_id: 'plan-demo',
        notes: 'Introduces the Release Manager module.',
        changes: [
          {
            change_id: 'c-001',
            change_type: 'feat',
            description: 'Add Release Manager module (Module 5)',
            task_id: 'us1-001',
            story_id: 'US-1',
          },
          {
            change_id: 'c-002',
            change_type: 'feat',
            description: 'Add changelog Markdown generation',
            task_id: 'us1-002',
            story_id: 'US-2',
          },
          {
            change_id: 'c-003',
            change_type: 'feat',
            description: 'Add release summary statistics endpoint',
            task_id: 'us2-001',
            story_id: 'US-3',
          },
          {
            change_id: 'c-004',
            change_type: 'docs',
            description: 'Document all Module 5 REST endpoints',
            task_id: null,
            story_id: null,
          },
          {
            change_id: 'c-005',
            change_type: 'test',
            description: 'Add 40 integration tests for releases API',
            task_id: null,
            story_id: null,
          },
        ],
      },
      {
        version: '4.0.0',
        title: 'Task Forge',
        date: '2026-02-15',
        status: 'published',
        task_list_id: null,
        spec_id: null,
        plan_id: null,
        notes: 'Introduces the Task Forge module.',
        changes: [
          {
            change_id: 'c-101',
            change_type: 'feat',
            description: 'Add Task Forge module (Module 4)',
            task_id: null,
            story_id: 'US-4',
          },
          {
            change_id: 'c-102',
            change_type: 'feat',
            description: 'Add dependency graph with cycle detection',
            task_id: null,
            story_id: 'US-5',
          },
          {
            change_id: 'c-103',
            change_type: 'feat',
            description: 'Add GitHub Issues export preview',
            task_id: null,
            story_id: 'US-6',
          },
          {
            change_id: 'c-104',
            change_type: 'fix',
            description: 'Fix phase ordering in YAML persistence',
            task_id: 'fix-001',
            story_id: null,
          },
        ],
      },
      {
        version: '3.0.0',
        title: 'Architecture Planner',
        date: '2026-01-10',
        status: 'published',
        task_list_id: null,
        spec_id: null,
        plan_id: null,
        notes: 'Introduces the Architecture Planner module.',
        changes: [
          {
            change_id: 'c-201',
            change_type: 'feat',
            description: 'Add Architecture Planner module (Module 3)',
            task_id: null,
            story_id: null,
          },
          {
            change_id: 'c-202',
            change_type: 'feat',
            description: 'Add ER diagram and tech stack selector',
            task_id: null,
            story_id: null,
          },
        ],
      },
      {
        version: '0.1.0',
        title: 'WIP: Next Steps',
        date: '',
        status: 'draft',
        task_list_id: null,
        spec_id: null,
        plan_id: null,
        notes: 'Upcoming features being planned.',
        changes: [
          {
            change_id: 'c-301',
            change_type: 'feat',
            description: 'Planned: CI/CD pipeline integration',
            task_id: null,
            story_id: null,
          },
          {
            change_id: 'c-302',
            change_type: 'breaking',
            description: 'Planned: Rename projects root env variable',
            task_id: null,
            story_id: null,
          },
        ],
      },
    ],
  };
}

File: src/specify_cli/releases.py
"""
Release Manager — Module 5 of SpecForge

Implements the data structures and logic for the Release Manager module,
which tracks versioned releases, changelog entries, and release lifecycle
status for a software project.

Key Entities:
- ChangeType     — enum: feat | fix | docs | refactor | test | chore | breaking
- ReleaseStatus  — enum: draft | published | yanked
- ChangeEntry    — a single changelog item linked to a task/story
- Release        — a versioned release with status and change entries
- ReleaseLog     — root aggregate; persisted as releases.yaml
"""

from __future__ import annotations

import uuid
from enum import Enum
from pathlib import Path
from typing import Optional

import yaml


# ---------------------------------------------------------------------------
# Enumerations
# ---------------------------------------------------------------------------


class ChangeType(str, Enum):
    FEAT = "feat"
    FIX = "fix"
    DOCS = "docs"
    REFACTOR = "refactor"
    TEST = "test"
    CHORE = "chore"
    BREAKING = "breaking"


class ReleaseStatus(str, Enum):
    DRAFT = "draft"
    PUBLISHED = "published"
    YANKED = "yanked"


# Canonical change type display order (breaking first, then feat, etc.)
CHANGE_TYPE_ORDER: list[ChangeType] = [
    ChangeType.BREAKING,
    ChangeType.FEAT,
    ChangeType.FIX,
    ChangeType.REFACTOR,
    ChangeType.DOCS,
    ChangeType.TEST,
    ChangeType.CHORE,
]

CHANGE_TYPE_LABEL: dict[ChangeType, str] = {
    ChangeType.BREAKING: "Breaking Changes",
    ChangeType.FEAT: "Features",
    ChangeType.FIX: "Bug Fixes",
    ChangeType.REFACTOR: "Refactoring",
    ChangeType.DOCS: "Documentation",
    ChangeType.TEST: "Tests",
    ChangeType.CHORE: "Chores",
}


# ---------------------------------------------------------------------------
# ChangeEntry — value object
# ---------------------------------------------------------------------------


class ChangeEntry:
    """A single changelog item.

    Attributes
    ----------
    change_id:   Unique identifier (UUID)
    change_type: Category of change (ChangeType)
    description: Human-readable description of the change
    task_id:     Optional reference to the Task that produced this change
    story_id:    Optional reference to the User Story driving this change
    """

    def __init__(
        self,
        change_type: ChangeType | str,
        description: str,
        *,
        change_id: Optional[str] = None,
        task_id: Optional[str] = None,
        story_id: Optional[str] = None,
    ) -> None:
        self.change_id = change_id or str(uuid.uuid4())
        self.change_type = ChangeType(change_type) if isinstance(change_type, str) else change_type
        self.description = description
        self.task_id = task_id
        self.story_id = story_id

    def to_dict(self) -> dict:
        return {
            "change_id": self.change_id,
            "change_type": self.change_type.value,
            "description": self.description,
            "task_id": self.task_id,
            "story_id": self.story_id,
        }

    @classmethod
    def from_dict(cls, d: dict) -> "ChangeEntry":
        return cls(
            change_id=d["change_id"],
            change_type=d["change_type"],
            description=d["description"],
            task_id=d.get("task_id"),
            story_id=d.get("story_id"),
        )


# ---------------------------------------------------------------------------
# Release entity
# ---------------------------------------------------------------------------


class Release:
    """A versioned software release.

    Attributes
    ----------
    version:       SemVer string, e.g. "1.2.0"
    title:         Optional human-readable title
    date:          ISO-8601 date string, e.g. "2026-04-03"
    status:        ReleaseStatus (draft | published | yanked)
    task_list_id:  Optional reference to a Module 4 TaskList
    spec_id:       Optional reference to a Module 2 Spec
    plan_id:       Optional reference to a Module 3 Plan
    notes:         Free-form release notes
    changes:       Ordered list of ChangeEntry items
    """

    def __init__(
        self,
        version: str,
        *,
        title: str = "",
        date: str = "",
        status: ReleaseStatus | str = ReleaseStatus.DRAFT,
        task_list_id: Optional[str] = None,
        spec_id: Optional[str] = None,
        plan_id: Optional[str] = None,
        notes: str = "",
    ) -> None:
        self.version = version
        self.title = title
        self.date = date
        self.status = ReleaseStatus(status) if isinstance(status, str) else status
        self.task_list_id = task_list_id
        self.spec_id = spec_id
        self.plan_id = plan_id
        self.notes = notes
        self._changes: list[ChangeEntry] = []

    # ------------------------------------------------------------------
    # Change management
    # ------------------------------------------------------------------

    def add_change(self, entry: ChangeEntry) -> None:
        """Append a change entry to this release."""
        self._changes.append(entry)

    def remove_change(self, change_id: str) -> bool:
        """Remove a change entry by ID; returns True if found and removed."""
        for i, entry in enumerate(self._changes):
            if entry.change_id == change_id:
                del self._changes[i]
                return True
        return False

    def get_change(self, change_id: str) -> Optional[ChangeEntry]:
        """Return the change entry with *change_id*, or None."""
        for entry in self._changes:
            if entry.change_id == change_id:
                return entry
        return None

    @property
    def changes(self) -> list[ChangeEntry]:
        return list(self._changes)

    # ------------------------------------------------------------------
    # Status transitions
    # ------------------------------------------------------------------

    def publish(self) -> None:
        """Transition status to published."""
        self.status = ReleaseStatus.PUBLISHED

    def yank(self) -> None:
        """Transition status to yanked."""
        self.status = ReleaseStatus.YANKED

    def revert_to_draft(self) -> None:
        """Transition status back to draft."""
        self.status = ReleaseStatus.DRAFT

    def update_status(self, new_status: ReleaseStatus | str) -> None:
        self.status = ReleaseStatus(new_status) if isinstance(new_status, str) else new_status

    # ------------------------------------------------------------------
    # Changelog rendering
    # ------------------------------------------------------------------

    def changelog_markdown(self) -> str:
        """Render this release as a Markdown changelog section."""
        heading = f"## [{self.version}]"
        if self.date:
            heading += f" — {self.date}"
        if self.title:
            heading += f" *{self.title}*"
        lines = [heading, ""]
        if self.notes:
            lines += [self.notes, ""]
        by_type: dict[ChangeType, list[ChangeEntry]] = {ct: [] for ct in CHANGE_TYPE_ORDER}
        for entry in self._changes:
            by_type[entry.change_type].append(entry)
        for ct in CHANGE_TYPE_ORDER:
            entries = by_type[ct]
            if entries:
                lines.append(f"### {CHANGE_TYPE_LABEL[ct]}")
                lines.append("")
                for e in entries:
                    lines.append(f"- {e.description}")
                lines.append("")
        return "\n".join(lines).rstrip() + "\n"

    # ------------------------------------------------------------------
    # Serialization
    # ------------------------------------------------------------------

    def to_dict(self) -> dict:
        return {
            "version": self.version,
            "title": self.title,
            "date": self.date,
            "status": self.status.value,
            "task_list_id": self.task_list_id,
            "spec_id": self.spec_id,
            "plan_id": self.plan_id,
            "notes": self.notes,
            "changes": [c.to_dict() for c in self._changes],
        }

    @classmethod
    def from_dict(cls, d: dict) -> "Release":
        release = cls(
            version=d["version"],
            title=d.get("title", ""),
            date=d.get("date", ""),
            status=d.get("status", ReleaseStatus.DRAFT),
            task_list_id=d.get("task_list_id"),
            spec_id=d.get("spec_id"),
            plan_id=d.get("plan_id"),
            notes=d.get("notes", ""),
        )
        for c in d.get("changes", []):
            release._changes.append(ChangeEntry.from_dict(c))
        return release


# ---------------------------------------------------------------------------
# ReleaseLog — root aggregate
# ---------------------------------------------------------------------------


class ReleaseLog:
    """Root aggregate for Module 5.

    Manages all releases for a project and exposes changelog generation
    and summary statistics.
    Persisted as ``releases.yaml`` inside the project directory.

    Attributes
    ----------
    project_name: Human-readable project name
    notes:        Free-form notes about the release log
    """

    def __init__(
        self,
        project_name: str,
        *,
        notes: str = "",
    ) -> None:
        self.project_name = project_name
        self.notes = notes
        self._releases: list[Release] = []

    # ------------------------------------------------------------------
    # Release management
    # ------------------------------------------------------------------

    def add_release(self, release: Release) -> None:
        """Add a new release; raises ValueError if version already exists."""
        if any(r.version == release.version for r in self._releases):
            raise ValueError(f"Release version '{release.version}' already exists")
        self._releases.append(release)

    def get_release(self, version: str) -> Optional[Release]:
        """Return the release with *version*, or None."""
        for r in self._releases:
            if r.version == version:
                return r
        return None

    def remove_release(self, version: str) -> bool:
        """Remove a release by version; returns True if found and removed."""
        for i, r in enumerate(self._releases):
            if r.version == version:
                del self._releases[i]
                return True
        return False

    @property
    def releases(self) -> list[Release]:
        """Releases in the order they were added (newest first by convention)."""
        return list(self._releases)

    # ------------------------------------------------------------------
    # Changelog
    # ------------------------------------------------------------------

    def changelog_markdown(self) -> str:
        """Render the full changelog for all releases as Markdown."""
        lines = [f"# Changelog — {self.project_name}", ""]
        if self.notes:
            lines += [self.notes, ""]
        for release in self._releases:
            lines.append(release.changelog_markdown())
        return "\n".join(lines)

    # ------------------------------------------------------------------
    # Summary
    # ------------------------------------------------------------------

    def summary(self) -> dict:
        """Return counts of releases by status and total change counts."""
        by_status: dict[str, int] = {s.value: 0 for s in ReleaseStatus}
        by_change_type: dict[str, int] = {ct.value: 0 for ct in ChangeType}
        total_changes = 0
        for release in self._releases:
            by_status[release.status.value] += 1
            for entry in release.changes:
                by_change_type[entry.change_type.value] += 1
                total_changes += 1
        return {
            "total_releases": len(self._releases),
            "by_status": by_status,
            "total_changes": total_changes,
            "by_change_type": by_change_type,
        }

    # ------------------------------------------------------------------
    # Persistence
    # ------------------------------------------------------------------

    def save(self, path: Path) -> None:
        """Persist this release log to *path* (YAML)."""
        path.parent.mkdir(parents=True, exist_ok=True)
        data = {
            "project_name": self.project_name,
            "notes": self.notes,
            "releases": [r.to_dict() for r in self._releases],
        }
        with path.open("w", encoding="utf-8") as fh:
            yaml.safe_dump(data, fh, allow_unicode=True, sort_keys=False)

    @classmethod
    def load(cls, path: Path) -> "ReleaseLog":
        """Load a ReleaseLog from a YAML file; raises FileNotFoundError if absent."""
        with path.open(encoding="utf-8") as fh:
            data = yaml.safe_load(fh)
        log = cls(
            project_name=data["project_name"],
            notes=data.get("notes", ""),
        )
        for release_data in data.get("releases", []):
            log._releases.append(Release.from_dict(release_data))
        return log
